### PR TITLE
None Overrides

### PR DESCRIPTION
**Pull request by moebrowne**
As per the [CSP spec](https://www.w3.org/TR/CSP3/#serialized-source-list) the value of `none` should only ever exist on it's own.

This PR changes the `addDirective` method to replace all values when `none` is added and for `none` to be removed when other values are added.

A use case for this is where you have a policy which extends a stricter policy and you want to override it:

```php
class BasePolicy extends Policy {
   public function configure()
    {
        $this->addDirective(Directive::IMG, [Keyword::NONE])
    }
}

class ChildPolicy extends BasePolicy {
   public function configure()
    {
        parent::configure();

        $this->addDirective(Directive::IMG, [Keyword::SELF])
    }
}
```

Before this would have created a CSP with the value `'none' 'self'` which, if I'm reading the [spec](https://www.w3.org/TR/CSP3/#serialized-source-list) correctly, isn't allowed. I'm not 100% sure what browsers do in this scenario but Firefox gives the slightly worrying message `Ignoring unknown option 'none'`

### CODE REVIEW

**Comment by mfn** [comment]
Are we fine with _silently_ removing the existing ones, when none is passed?

**Comment by Gummibeer** [comment]
@mfn I'm fine with it. Would you expect an exception if one of the `none` conflicts comes up?

**Comment by mfn** [comment]
It's something I crossed my mind.

But the problem now is: you can some default configured object and only want to override a specific part and 💥 , you really can't.

Sorry no better feedback 🤷‍♀️

**Comment by moebrowne** [comment]
I could get on board with an exception being thrown if you try to set `none` with another value at the same time. Eg:

```
$this->addDirective(Directive::CONNECT, [Keyword::NONE, 'connect']);
```

but not if you're explicitly saying set the directive to `none`

**Comment by mfn** [comment]
> I could get on board with an exception being thrown if you try to set `none` with another value at the same time. Eg:

I like the idea 👍

**Comment by moebrowne** [comment]
The new exception has been implemented

File: src/Policies/Policy.php [hunk 1]
```diff
@@ -42,6 +45,16 @@ public function addDirective(string $directive, $values): self
             return explode(' ', $value);
         }, Arr::wrap($values))));
 
+        if (in_array(Keyword::NONE, $values, true) === true) {
+            $this->directives[$directive] = [$this->sanitizeValue(Keyword::NONE)];
+
+            return $this;
+        } else {
```
Comment (Gummibeer) [inline src/Policies/Policy.php]: `else` is not necessary with an early return.

File: src/Policies/Policy.php [hunk 2]
```diff
@@ -117,6 +130,13 @@ protected function guardAgainstInvalidDirectives(string $directive)
         }
     }
 
+    protected function guardAgainstInvalidValues(array $values)
+    {
+        if (in_array(Keyword::NONE, $values, true) === true && count($values) > 1) {
```
[the remark this answers is not in the document]
Comment (moebrowne) [reply src/Policies/Policy.php]: I prefer to use strict conditions but that's just personal preference

File: src/Policies/Policy.php (hunk 1 above)
Comment (moebrowne) [reply src/Policies/Policy.php]: True, I had noticed this but thought it made the intent more obvious

**Comment by Gummibeer** [comment]
poke @spatie @freekmurze 
I think that this PR is still valid

**Pull request merged**